File: java/client-apis/src/main/java/org/apache/rocketmq/client/apis/ClientServiceProvider.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.rocketmq.client.apis;

import java.util.Iterator;
import java.util.ServiceLoader;
import org.apache.rocketmq.client.apis.consumer.LitePushConsumerBuilder;
import org.apache.rocketmq.client.apis.consumer.LiteSimpleConsumerBuilder;
import org.apache.rocketmq.client.apis.consumer.PushConsumerBuilder;
import org.apache.rocketmq.client.apis.consumer.SimpleConsumerBuilder;
import org.apache.rocketmq.client.apis.message.MessageBuilder;
import org.apache.rocketmq.client.apis.producer.ProducerBuilder;

/**
 * Service provider to seek client, which load client according to
 * <a href="https://en.wikipedia.org/wiki/Service_provider_interface">Java SPI mechanism</a>.
 */
public interface ClientServiceProvider {

    /**
     * To avoid potential concurrency issues, the {@link #loadService()} logic
     * has been changed to use lazy initialization with caching:
     * <p>
     * 1. Lazy loading + caching:
     * - On the first call, the implementation is loaded via {@link ServiceLoader}
     * and cached in {@link Holder#INSTANCE};
     * - Subsequent calls simply return the cached instance.
     * <p>
     * 2. If you need the old behavior (i.e., always load through ServiceLoader
     * each time), you can call {@link #doLoad()} directly:
     * - {@link #doLoad()} does not cache anything; it creates a new ServiceLoader
     * and loads an implementation on every call;
     * - You are responsible for handling any concurrency control when using
     * {@link #doLoad()} directly.
     */

    class Holder {
        static volatile ClientServiceProvider INSTANCE;

        private Holder() {
            // prevents instantiation
        }
    }

    static ClientServiceProvider loadService() {
        ClientServiceProvider inst = Holder.INSTANCE;
        if (inst != null) {
            return inst;
        }
        synchronized (ClientServiceProvider.class) {
            if (Holder.INSTANCE != null) {
                return Holder.INSTANCE;
            }
            Holder.INSTANCE = doLoad();
            return Holder.INSTANCE;
        }
    }

    static ClientServiceProvider doLoad() {
        final ServiceLoader<ClientServiceProvider> loaders = ServiceLoader.load(ClientServiceProvider.class);
        final Iterator<ClientServiceProvider> iterators = loaders.iterator();
        if (iterators.hasNext()) {
            return iterators.next();
        }
        throw new UnsupportedOperationException("Client service provider not found");
    }

    /**
     * Get the producer builder by the current provider.
     *
     * @return the producer builder instance.
     */
    ProducerBuilder newProducerBuilder();

    /**
     * Get the message builder by the current provider.
     *
     * @return the message builder instance.
     */
    MessageBuilder newMessageBuilder();

    /**
     * Get the push consumer builder by the current provider.
     *
     * @return the push consumer builder instance.
     */
    PushConsumerBuilder newPushConsumerBuilder();

    /**
     * Get the lite push consumer builder by the current provider.
     *
     * @return the lite push consumer builder instance.
     */
    LitePushConsumerBuilder newLitePushConsumerBuilder();

    /**
     * Get the lite simple consumer builder by the current provider.
     *
     * @return the lite simple consumer builder instance.
     */
    LiteSimpleConsumerBuilder newLiteSimpleConsumerBuilder();

    /**
     * Get the simple consumer builder by the current provider.
     *
     * @return the simple consumer builder instance.
     */
    SimpleConsumerBuilder newSimpleConsumerBuilder();
}

File: php/ClientServiceProvider.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

namespace Apache\Rocketmq;

use Apache\Rocketmq\Builder\LiteProducerBuilder;
use Apache\Rocketmq\Builder\LitePushConsumerBuilder;
use Apache\Rocketmq\Builder\LiteSimpleConsumerBuilder;
use Apache\Rocketmq\Builder\MessageBuilder;
use Apache\Rocketmq\Builder\ProducerBuilder;
use Apache\Rocketmq\Builder\PushConsumerBuilder;
use Apache\Rocketmq\Builder\SimpleConsumerBuilder;

/**
 * Service provider interface for RocketMQ client components.
 * 
 * This interface provides a unified entry point for creating various RocketMQ client components
 * using the Builder pattern. It follows the Service Provider Interface (SPI) design pattern,
 * similar to Java's ClientServiceProvider.
 * 
 * Usage example:
 * ```php
 * $provider = ClientServiceProvider::loadService();
 * 
 * // Create producer
 * $producer = $provider->newProducerBuilder()
 *     ->setClientConfiguration($config)
 *     ->setTopics('my-topic')
 *     ->build();
 * 
 * // Create push consumer
 * $consumer = $provider->newPushConsumerBuilder()
 *     ->setClientConfiguration($config)
 *     ->setConsumerGroup('my-group')
 *     ->setTopic('my-topic')
 *     ->setMessageListener($listener)
 *     ->build();
 * ```
 */
interface ClientServiceProvider
{
    /**
     * Get a new producer builder instance.
     *
     * @return ProducerBuilder The producer builder instance
     */
    public function newProducerBuilder(): ProducerBuilder;

    /**
     * Get a new message builder instance.
     *
     * @return MessageBuilder The message builder instance
     */
    public function newMessageBuilder(): MessageBuilder;

    /**
     * Get a new push consumer builder instance.
     *
     * @return PushConsumerBuilder The push consumer builder instance
     */
    public function newPushConsumerBuilder(): PushConsumerBuilder;

    /**
     * Get a new lite push consumer builder instance.
     *
     * @return LitePushConsumerBuilder The lite push consumer builder instance
     */
    public function newLitePushConsumerBuilder(): LitePushConsumerBuilder;

    /**
     * Get a new lite simple consumer builder instance.
     *
     * @return LiteSimpleConsumerBuilder The lite simple consumer builder instance
     */
    public function newLiteSimpleConsumerBuilder(): LiteSimpleConsumerBuilder;

    /**
     * Get a new simple consumer builder instance.
     *
     * @return SimpleConsumerBuilder The simple consumer builder instance
     */
    public function newSimpleConsumerBuilder(): SimpleConsumerBuilder;

    /**
     * Get a new lite producer builder instance.
     *
     * @return LiteProducerBuilder The lite producer builder instance
     */
    public function newLiteProducerBuilder(): LiteProducerBuilder;

    /**
     * Load the default service provider implementation.
     * 
     * This method uses lazy initialization with caching to avoid potential
     * concurrency issues. On the first call, it creates and caches the default
     * provider instance. Subsequent calls return the cached instance.
     *
     * @return ClientServiceProvider The service provider instance
     */
    public static function loadService(): ClientServiceProvider;
}

// Load the default implementation
require_once __DIR__ . '/DefaultClientServiceProvider.php';
